exec/java-exec/src/main/java/org/apache/drill/exec/store/log/LogRecordReader.java

```diff
@@ -0,0 +1,623 @@
+        return new VarCharDefn(name, patternIndex);
+      case INT:
+      case SMALLINT:
+      case BIGINT:
+        return new IntDefn(name, patternIndex);
+      case FLOAT8:
+      case FLOAT4:
+      case DECIMAL9:
+      case DECIMAL18:
+        return new DoubleDefn(name, patternIndex);
+      case DATE:
+        return new DateDefn(name, patternIndex, formatConfig.getDateFormat());
+      case TIMESTAMP:
+        return new TimeStampDefn(name, patternIndex, formatConfig.getDateFormat());
+      case TIME:
+        return new TimeDefn(name, patternIndex, formatConfig.getTimeFormat());
+      default:
+        throw UserException
+            .validationError()
+            .message("Undefined column types")
+            .addContext("Position", patternIndex)
+            .addContext("Field name", name)
+            .addContext("Type", typeName)
+            .build(logger);
+    }
+  }
+
+  private void defineVectors() {
+    for (int i = 0; i < columns.length; i++) {
+      try {
+        columns[i].define(outputMutator);
+      } catch (SchemaChangeException e) {
+        throw UserException
+            .systemError(e)
+            .message("Vector creation failed")
+            .build(logger);
+      }
+    }
+  }
+
+  @Override
+  public int next() {
+    rowIndex = 0;
+    while (nextLine()) {
+    }
+    return rowIndex;
+  }
+
+  private boolean nextLine() {
+    String line;
+    try {
+      line = reader.readLine();
+    } catch (IOException e) {
+      throw UserException
+          .dataReadError(e)
+          .message("Error reading file:")
+          .addContext("File", fileWork.getPath())
+          .build(logger);
+    }
+    if (line == null) {
+      return false;
+    }
+    Matcher m = pattern.matcher(line);
+    if (m.matches()) {
+      loadVectors(m);
+      return rowIndex < BATCH_SIZE;
+    }
+    errorCount++;
+    if (errorCount < formatConfig.getMaxErrors()) {
+      logger.warn("Unmatached line: {}", line);
+    } else if( errorCount > formatConfig.getMaxErrors() ){
+      throw UserException.parseError()
+          .message("Too many errors.  Max error threshold exceeded.")
+          .addContext("Line", line)
+          .addContext("Line number", rowIndex)
+          .build(logger);
+    }
+    //If the user asked for the unmatched columns display them
+    if(unmatchedColumnIndex != -1 ) {
+      columns[unmatchedColumnIndex].load(rowIndex, line);
+      rowIndex++;
+      return rowIndex < BATCH_SIZE;
+    }
+
+
+    return true;
+  }
+
+  private void loadVectors(Matcher m) {
+    for (int i = 0; i < columns.length; i++) {
+      if (columns[i].index >= 0) {
+        String value = m.group(columns[i].index + 1);
+        if (value != null) {
+          columns[i].load(rowIndex, value);
+        }
+      } else if( columns[i].index == -1 ){
+        String value = m.group(0);
+        if (value != null) {
+          columns[i].load(rowIndex, value);
+        } else {
+          rowIndex++;
```

Comment:
Ah oh... The row index is incremented twice. Is that what was wanted? Returning two rows, one with values filled in, one with all nulls?

The above issues raise an important question: do unit tests exist for all these cases? If so, the tests should have found the problems, not your code reviewer.

Reply:
So the way it works, is that if you have a data file with 5 rows, 4 of which match, and you do a SELECT *, you get 4 rows back.   If you do a `SELECT _raw` you also get 4 rows back.  
Right now, if you do a `SELECT _unmatched_rows` you get four null rows and 1 row with the unmatched row. 

I'm not sure what the best behavior is honestly, but I was trying to balance usability in that we don't want Drill to break every time it encounters bad data, but at the same time, we also want the user to have the ability to identify rows that don't match the regex.  I really like the idea of the `max_errors` parameter, in that you can tune how sensitive you want Drill to be.  If you want Drill to blow up every time it encounters an error you can set that to zero.